### PR TITLE
skip uncompress if header indicates length 0

### PR DESCRIPTION
**Pull request by flushentitypacket**

### CODE REVIEW

File: header.go [hunk 1]
```diff
@@ -60,6 +60,9 @@ func UncompressAllocHdr(out, in []byte) ([]byte, error) {
 		return out, errTooShort
 	}
 	origlen := binary.LittleEndian.Uint32(in)
+	if origlen <= 0 {
+		return out, nil
+	}
```
Comment (HippoBaro) [inline header.go]: Have you tested this? Looking at the control flow of the function, this will never trigger: the function would return `errTooShort`.

**Review by HippoBaro — CHANGES_REQUESTED** [review]

File: header.go (hunk 1 above)
Comment (flushentitypacket) [reply header.go]: This is for if there is a header whose value is 0. So there are 4 bytes present for the header, but its uint32 value is 0.

Comment (HippoBaro) [reply header.go]: Oh, I see!. It seems to me like this should return `errTooShort` or another error then, wdyt?

Comment (flushentitypacket) [reply header.go]: I _think_ this is a valid case where this is a payload for which there is null content, so I don't think an error is appropriate. Just returning the null content back makes more sense to me

**Review by HippoBaro — APPROVED** [review]

File: header_test.go [hunk 2]
```diff
@@ -128,6 +128,31 @@ func TestCompressAllocHdr(t *testing.T) {
...
+	input := []byte("")
+	output := make([]byte, CompressBoundHdr(input))
+	outSize, err := CompressHdr(output, input)
+	if err != nil {
+		t.Fatalf("Compression failed: %v", err)
+	}
+	if outSize == 0 {
+		t.Fatal("Output buffer is empty.")
+	}
+	// This is the key line. Some lz4 implementations don't trim their compressed outputs for blank payloads,
```
Comment (evanj) [inline header_test.go:141]: Oh wow, does this mean we actually have a bug for other inputs also? Like what happens if a non-empty compressed payload has extra data? Or what is UncompressAllocHdr even supposed to do in these cases? :) The doc comment certainly does not specify...

One moment let me test this ...

File: header.go (hunk 1 above)
Comment (evanj) [reply header.go]: For clarity, without this change, `UncompressAllocHdr` returns an error: `Decompression failed: Malformed compression stream` I think this is arguably the correct result? We wrap `LZ4_decompress_safe` with these functions, and the error coms from it. Its documentation says:

```
compressedSize : is the exact complete size of the compressed block.
...
If the source stream is detected malformed, the function will stop decoding and return a negative result.
This function is protected against malicious data packets.
```

https://github.com/lz4/lz4/blob/v1.8.3/lib/lz4.h#L141-L148

So this error happens because a caller passed a buffer that is too long to `UncompressAllocHdr`.

Additionally: LZ4 produces output for compressing this message. The buffer contents are: 5 `0x00` bytes. Passing anything other than those 5 bytes should be an error?

The same problem happens with other inputs. E.g. if I pass in `"abc"`, LZ4 outputs 8 bytes. If I pass 9 bytes into UncompressAllocHdr, we get an error. I think that is correct?

Maybe another way to approach this: in what context did this error happen? Or maybe we should clarify the requirements about how to use these functions on the doc comments?

Comment (flushentitypacket) [reply header.go]: Fair points for sure. For context, this came up when decompressing lz4 payloads from dogweb, which seems like behaves a bit differently than this impl when it comes to empty inputs. And yeah, Malformed compression stream is the error that we get.

Comment (flushentitypacket) [reply header.go]: Also, I'm not actually sure what makes sense in this empty payload case anyway. Compressing a zero-length message should arguably be 4 `0x00` bytes instead of 5, right? But AFAIK we're not following any sort of actual standard here with the header format--just a de-facto standard.

EDIT: Actually, 5 makes sense given that `0x00` is the terminator.

Comment (flushentitypacket) [reply header.go]: You can see the root of the problem in this PR which fixes it: https://github.com/DataDog/python-lz4/pull/1/files

TLDR is the python-lz4 lib leaves garbage bytes at the end of its output when given a zero-length input.

Now I think there's a debate whether it makes sense to fix this with the python-lz4 fork, which would be a new fork, or to just fix it here.

More context and discussion here: https://github.com/DataDog/dogweb/pull/73067

**Comment by flushentitypacket** [comment]
Closed in favor of https://github.com/DataDog/dogweb/pull/73698
